### DISCUSSION TITLE
How to get last n ancestors as revset

### DISCUSSION THREAD

**Discussion by sullyj3** [jj-vcs/jj#1984] [Q&A] [answered]
I'm trying to figure out how to do the equivalent of `git log -n 5`. I came up with `jj log -r @-----::@`. Is there a better way? (This method clearly doesn't scale)

**Comment by martinvonz** [comment, marked as answer]
The revset engine already knows how to do it, so I think it's just a matter of updating the `ancestors()` function to accept a generation range (@yuja can correct me if I'm wrong). That code is here: https://github.com/martinvonz/jj/blob/b19bf3757fe816da4f6079e388c8f9abb34ae214/lib/src/revset.rs#L946-L950

I'm not sure exactly what the syntax should be. Mercurial has `ancestors(x, n)` to give up to n generations up to the specified distance. I think it makes sense to copy that. We could also accept `ancestors(x, m, n)` to also allow a lower bound on the number of generation (i.e. making `ancestors(x, n)` equal to `ancestors(x, 0, n)`).

There's also support in the revset engine for descendants in a specified range, so it shouldn't be hard to add syntax for that too.

Do you feel like implementing any of that?

**Reply by sullyj3** [reply]
I'll have a read through of the code and see if I feel capable of tackling it

**Reply by yuja** [reply]
Yep, it's just a matter of syntax. Abusing range expression `ancestors(x, m..n)` or optional `start=m`/`end=n` parameters `ancestors(x, m, n)`.

**Reply by sullyj3** [reply]
PR at #2200
